### ISSUE TITLE
Invoke-WebRequest should support -Encoding parameter for default character set

### ISSUE THREAD

**Issue by rhubarb-geek-nz** [PowerShell/PowerShell#19226]
### Summary of the new feature / enhancement

This is a follow up to https://github.com/PowerShell/PowerShell/issues/19156

Other cmdlets use an "-Encoding" parameter, eg Get-Content. It would be nice if this was used for the default encoding character set when the character set is not present in the ContentType and the body is some form of string rather than binary. This would be better than treating certain MIME types as somehow special.

### Proposed technical implementation details (optional)

The Invoke-WebRequest refers to a default character set but it is not clear how this is determined, and the behaviour is different between 7.3.3 and powershell-7.4.0-preview.1

The concept is to use the "-Encoding" parameter to specify the default character set if it is not present in the MIME type.

An example is "application/json" is technically a binary encoding so should not have the charset in the ContentType. 

It would be better to use an explicit "-Encoding" parameter than maintain a list of special MIME types that need particular character set encodings.

**Comment by mklement0** [comment]
That would be handy.

For context: 

* The encoding PowerShell currently uses for a `[string]` `-Body` argument is inferred from the `charset` value of a `-ContentType` argument, e.g. `-ContentType 'application/json; charset=utf8'`
  * That it _isn't_ also inferred from an equivalent `'Content-Type'` entry in a hashtable passed to `-Headers` is the subject of https://github.com/PowerShell/PowerShell/issues/19156

* Since both methods are somewhat cumbersome, and for consistence with existing cmdlets that support `-Encoding`, it would be nice to able to specify the request encoding for `[string]` -Body arguments via `-Encoding`.

* I don't think `application/json` is a _binary_ encoding; it is clearly _text_-based and its prescribed default is UTF-8; from [RFC 4627](https://www.ietf.org/rfc/rfc4627.txt):
  * > JSON text SHALL be encoded in Unicode.  The default encoding is
   UTF-8.
  * > JSON may be represented using UTF-8, UTF-16, or UTF-32.  When JSON
      is written in UTF-8, JSON is 8bit compatible.  When JSON is
      written in UTF-16 or UTF-32, the binary content-transfer-encoding
      must be used. 

---

Given the above:

* I do think that `charset` should be present (unless it is the default for a given MIME type, but that would require the web cmdlets to have that knowledge), and an `-Encoding` argument should automatically be _translated into one_. Also, given that some text-based formats - including JSON - support _multiple_ character encodings, a deviation from the default must be specified. 
  * In other words: `-Encoding iso-8859-1 -Content-Type application/json` would just be a more PowerShell-friendly alternative to `-Content-Type 'application/json; charset=iso-8859-1'`
  * I presume that the `.WebName` property value of a given `[System.Text.Encoding]` instance is the right one to use; e.g., `[Text.Encoding]::GetEncoding(28591).WebName`

* As a result, either precedence rules must  be defined - e.g. `-Encoding` > `-ContentType` > `'Content-Type'` header field - or an error must be reported in case of conflicts.

* Separately, it would certainly help if the web cmdlets were aware of the default encoding of at least the most widely text-based MIME formats and apply those defaults automatically - I don't know how practical that is, but perhaps - with the world converging on UTF-8 - a general UTF-8 default is sufficient.

**Comment by rhubarb-geek-nz** [comment]
[Deprecated APPLICATION_JSON_UTF8_VALUE](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/MediaType.html#APPLICATION_JSON_UTF8_VALUE)

Springboot deprecated the application/json with charset utf-8 because it is redundant. Now it is just "application/json"

If the charset where automatically appended then in order to get just "application/json" then one would have to revert to binary encoding with a byte array like now, so defeating the purpose of this change.

IANA definitely do say binary, https://www.iana.org/assignments/media-types/application/json

**Comment by CarloToso** [comment]
Once #19227 is merged I can work on adding the `-Encoding` parameter, but I think we need the opinion of WG-Cmdlets-Utility before adding a new parameter to WebCmdlets @iSazonov @SteveL-MSFT

**Comment by mklement0** [comment]
Thanks, @rhubarb-geek-nz, you're right: 

* `application/json` _needs no_ `charset` attribute, and. per the [media-type definition](https://www.iana.org/assignments/media-types/application/json) you linked to:

  * > Note: No "charset" parameter is defined for this registration.
   Adding one really has no effect on compliant recipients.

And a bit of conceptual fog just lifted for me:

* The _binary_  in the media-type definition refers to the _transport_ encoding, not the _character_ encoding, which in the case of `application/json` is _limited to_ UTF-8 (default), UTF-16LE/BE, and UTF-32LE/BE:

  * > When JSON is written in UTF-8, JSON is 8bit compatible. 
  When JSON is written in UTF-16 or UTF-32, the binary content-transfer-encoding must be used.

  * Honestly, I don't how binary content-transfer-encoding comes into play with respect to PowerShell's web cmdlets: Is it automatically chosen behind the scenes when `[byte[]]` is passed to `-Body`?

----

Therefore:

* With `application/json`, specifically, `charset` is never _needed_, but seemingly also does _no harm_.

* Other media types (formerly known as MIME types), such as [`text/xml`](https://www.iana.org/assignments/media-types/text/xml) _do_ support a `charset` attribute, and require it when an encoding other than the default is used.

For backward compatibility alone we should continue to allow `charset` with `application/json` (so that _PowerShell_ knows what encoding to apply when it finds it in `-ContentType` or (soon) in a `Content-Type` header entry), but also because we don't want to prevent users from calling non-standard JSON web services that choose to implement `charset` support so as to support _additional_ encodings.

For consistency it therefore also makes sense to translate `-Encoding` into its equivalent `charset` attribute.
_Assuming we  have this knowledge built in_, this translation to `charset` can be _omitted_ if the target encoding is _known_ to be the target media type's default. But, I suspect that it's fine to just _categorically_ include whenever `-Encoding` is used.

**Comment by SteveL-MSFT** [comment]
I've queued this up to the WG

**Comment by rhubarb-geek-nz** [comment]
> But, I suspect that it's fine to just _categorically_ include it whenever `-Encoding` is used.

I would prefer to leave the user in control and not try and be too helpful/clever. If the user has specified the content type either as a parameter or a header we should trust that that is what they want and not append the charset. Invoke-WebRequest has curl as an alias so we should compare it with that tool. It does what you ask it to do rather than try and second-guess what you want.

**Comment by iSazonov** [comment]
> I would prefer to leave the user in control and not try and be too helpful/clever.

Sometimes .Net implementation could block this to build correct header. Needs more investigations.

**Comment by rhubarb-geek-nz** [comment]
> > I would prefer to leave the user in control and not try and be too helpful/clever.
> 
> Sometimes .Net implementation could block this to build correct header. Needs more investigations.

I would find that acceptable, if .Net rejects what I am trying to send using Invoke-WebRequest that is reasonable since I know that PowerShell is built on .Net and I would get an error. If Invoke-WeRequest silently changes the content type and I am unaware then that is not doing what I requested.

**Comment by mklement0** [comment]
@iSazonov

Note that [the docs](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-webrequest?view=powershell-7.3&WT.mc_id=ps-gethelp#-body) already state at least one case where `-ContentType` and `-Headers` get ignored: 
> When a MultipartFormDataContent object is supplied for Body, any Content related headers supplied to the ContentType, Headers, or WebSession parameters is overridden by the Content headers of the MultipartFormDataContent object. 

Are you thinking of additional cases, and, if so, do you have any pointers?

**Comment by mklement0** [comment]
@rhubarb-geek-nz 

>  `Invoke-WebRequest` has `curl` as an alias

* That is - fortunately - no longer true in PowerShell Core, and, to me, has always been a bad idea, as have all aliases been that were named for well-known external programs / standard utilities on other platforms (`curl`, `ls`) or `cmd.exe`'s internal commands (`dir`). The fundamentally different syntax and abilities of the utilities / internal commands alluded to makes these aliases marginally useful at best - in all but the simplest invocations - and, more typically a source of confusion / a nuisance (shadowing utilities such as `curl.exe` in Window PowerShell).

* `curl` is clearly a lower-level tool (you won't find the term `charset` in its `man` page), and whoever needs that, can now use it even on Windows, given that recent versions ship with `curl.exe`

> be too helpful/clever

Being helpful - read: provide a high level of abstraction - is one of PowerShell's core strengths.

In fact, it is pointless to introduce an `-Encoding` parameter _unless_ its argument gets translated into a `charset` `Content-Type` attribute.

Otherwise, you'll have to specify the encoding _twice_ in cases where you need it, which is _most_ text-based formats (JSON-based formats being a notable exception), say `text/plain` and `text/xml`, which defeats the purpose.

That is, you'd have to pass the desired encoding to `-Encoding` and _again_ as part of `-ContentType` (or `-Headers`), which isn't only duplicated effort but makes conflicts possible.

**Comment by iSazonov** [comment]
> Are you thinking of additional cases, and, if so, do you have any pointers?

I remember we had discussion about ContentLength.
There can be an infinite number of such situations.

**Comment by iSazonov** [comment]
.Net implementation can rebuild the header or block variants for various reasons.

**Comment by rhubarb-geek-nz** [comment]
I agree that `curl` and PowerShell's `Invoke-WebRequest` are different tools but this is also related to PowerShell dealing with objects where as UNIX shells deal with byte streams.

Now HTTP is essentially a binary transport mechanism, its purpose is to exchange data as bytes with further encoding described by the `Content-Type` header. Regarding the `Content-Length` header, the original HTTP did not even require you to use headers, the verb and url was in the first line, as was the response status code. The actual length of the data could be identified by the end-of-file indicator. This changed with the requirements of Keep-Alive.

> In fact, it is pointless to introduce an `-Encoding` parameter _unless_ its argument gets translated into a `charset` `Content-Type` attribute.

This is not case and the original intention was to define a fallback default encoding for strings given the version up to 7.3.3 was ambiguous about what default character encoding it actually used.

If I specify a charset encoding in the `Content-Type`, then all is well and `Invoke-WebRequest` knows how to translate the body if it is a string.

If I specify no `charset` encoding in the `Content-Type` and the body was string, then it should encode it using the  `charset` given in `Encoding`, and not change the `Content-Type` parameter or passed in `Header` values.

If the `Body` is a byte array or byte stream then the tool should use the given `Content-Type` and headers unchanged and use the given byte data unchanged.

The role of `Invoke-WebRequest` should still be like `curl`, to perform the HTTP requests and responses but to do it in a PowerShell environment. That does not mean it is somehow in a better position to know what I want to actually send other than what I have explicitly told it to do.

In terms of division of responsibility, `Invoke-WebRequest` should know how to make HTTP requests. The tools that should know about JSON should be `ConvertTo-Json` and `ConvertFrom-Json`. `Invoke-WebRequest` should not need any special coding for any of the currently 2884 known MIME types, or the unknown ones.

I gave the example of `SpringBoot` just using `application/json` as the `Content-Type` as it defaults its encoding to `UTF-8`. Its justification is that Chrome and other major browsers support this. Wanting to make request consistent with how Chrome makes requests and compatibility with how Chrome encodes JSON requests should not be considered an odd thing to want to or something that is deliberately prevented by the tool.

Likewise JSON:API. The [specification for JSON:API](https://jsonapi.org/format/) says

```
The JSON:API media type is application/vnd.api+json.

For example, in the header Content-Type: text/html; charset="utf-8", the media type is text/html and charset is a parameter.

The JSON:API media type MUST NOT be specified with any media type parameters other than ext and profile
```

So `charset` is an example of a parameter, but it MUST NOT be used with JSON:API

The [JSON API Examples](https://jsonapi.org/examples/) all show the Content-Type without charset parameter.

Eg

```
PATCH /posts/1 HTTP/1.1
Content-Type: application/vnd.api+json
Accept: application/vnd.api+json

{ "datum": [ ] }
```

In summary,

- The purpose of the `-Encoding` was supposed to be a replacement for the fallback default because it was ambiguous. 
- If the encoding exists in the Content-Type then that should be used, like in 7.3.3.
- If a Content-Type or Header is explicitly given it should be transmitted unchanged exactly as in 7.3.3.
- If the body is already byte array or byte stream then no conversion is done.
- The encoding is only used if the Body is some form of string or character data.

**Comment by rhubarb-geek-nz** [comment]
Same applies to [JSON-RPC over HTTP ](http://www.simple-is-better.org/json-rpc/transport_http.html)

A HTTP POST request MUST specify the following headers:

- Content-Type: MUST be application/json.
- Content-Length: MUST contain the correct length according to the HTTP-specification.
- Accept: MUST be application/json.

**Comment by mklement0** [comment]
I appreciate the thoughtful response, @rhubarb-geek-nz.

> The purpose of the -Encoding was supposed to be a replacement for the fallback default because it was ambiguous.

I see that we had different use cases in mind.
Before I address how your comments relate to the use case I had in mind, let's focus on yours:

* v7.4+ will default to UTF-8 for encoding `[string]`-based `-Body` arguments.

* _Older_ versions that do _not_ want to use a `charset` attribute will need to use something like the following, providing a `byte[]` array obtained via the `.GetBytes()` method of the character encoding of interest; _this is what will be required in cross-version, cross-edition code_:

```powershell
# Use UTF-8 encoding, passing a byte[] representation of it.
# Works in v7.3- too
(Invoke-RestMethod https://httpbin.org/post -Body ([Text.Utf8Encoding]::new().GetBytes('{ "foo": "sß" }')) -Method POST -ContentType application/json).json
```

* Presumably, your proposed `-Encoding`  parameter would be introduced in v7.4 at the earliest (and introducing it earlier wouldn't make sense, given the pending v7.4 change of the default character encoding).

In light of the above:

* What purpose would `-Encoding` serve?
* It won't help in _older_ versions, and it won't be necessary in v7.4+
* If the intent is to support exotic media types with non-UTF8 defaults / ad-hoc overrides of media types for non-standard web-service implementations: that doesn't call for a dedicated new parameter - just use a `byte[]`-typed `-Body` argument or - if the media type supports a `charset` attribute - use `-Content-Type`  with a `charset` attribute, as already supported.

**Comment by rhubarb-geek-nz** [comment]
@mklement0 Agreed, the issue originated after testing and found the implementation was ambiguous regarding the default character set, if it is always UTF-8 then it is consistent now with other parts of Dotnet Core.

If using a body of byte array or byte stream means that the headers and the content-type are not modified then that will permit the supporting of plain "application/json" and "application/vnd.api+json" which will permit my use cases.

Given that is consistent with existing behaviour it means that that strategy is also backwards compatible.

So if that is true then I am happy to withdraw the proposal.
